python/lsst/afw/image/multiband.py

```diff
@@ -0,0 +1,346 @@
+                self._filters = singles.filters
+                self._singles = singles.singles
+            else:
+                self._filters = [f for f in singles.filters]
+                self._singles = self._copySingles(singles)
+            if slices is None:
+                self._array = singles._array
+            else:
+                try:
+                    # This will fail if the first slice is a string,
+                    # or list of strings
+                    self._array = singles._array[slices]
+                    if len(singles.filters) > 1:
+                        if isinstance(slices[0], slice):
+                            self._filters = singles.filters[slices[0]]
+                        elif np.issubdtype(type(slices[0]), np.integer):
+                            self._filters = [singles.filters[slices[0]]]
+                        else:
+                            self._filters = [singles.filters[f] for f in slices[0]]
+                    else:
+                        self._filters = singles.filters
+                except IndexError:
+                    # Filter using the list of strings
+                    _filters = slices[0]
+                    slices = list(slices)
+                    slices[0] = [self.filters.index(f) for f in slices[0]]
+                    self._array = singles._array[slices]
+                    self._filters = _filters
+            # Set the bbox size based on the slices
+            x0 = singles.x0
+            y0 = singles.y0
+            if len(slices) > len(singles.array.shape):
+                err = "Too many indices, expected {0} but received {1}"
+                raise IndexError(err.format(len(singles.array.shape), len(slices)))
+            if len(slices) == len(singles.array.shape):
+                xslice = slices[-1]
+                if isinstance(xslice, slice):
+                    if xslice.start is not None:
+                        x0 += xslice.start
+                else:
+                    raise IndexError("x index must be a slice")
+            if len(slices) >= len(singles.array.shape)-1:
+                if len(slices) == len(singles.array.shape):
+                    yslice = slices[-2]
+                else:
+                    yslice = slices[-1]
+                if isinstance(yslice, slice):
+                    if yslice.start is not None:
+                        y0 += yslice.start
+                else:
+                    raise IndexError("y index must be a slice")
+            xy0 = Point2I(x0, y0)
+            xyF = Point2I(x0+self.array.shape[-1]-1, y0+self.array.shape[-2]-1)
+            self._xy0 = xy0
+            self._bbox = Box2I(xy0, xyF)
+            for singleObj in self.singles:
+                singleObj.setXY0(xy0)
+
+            # Create instances of the single band objects
+            # whose data points to the appropriate slice
+            # of self.array
+            self._updateSingles(singles.imageType)
+        else:
```

Comment:
I'd recommend splitting this up into at least two private methods called by `__init__` for clarity.